src/merge.rs

```diff
@@ -1796,62 +1788,98 @@ impl<'t> MergedRoot<'t> {
+                    // This is the case for invalid remote-only items. Upload a
+                    // tombstone, but don't flag it as remotely merged; we'll do
+                    // that when we store the uploaded tombstone.
+                    ops.insert_local_tombstones
+                        .push(InsertLocalTombstone(remote_node));
+                    ops.upload_tombstones
+                        .push(UploadTombstone(&remote_node.item().guid));
+                }
+                (Record::Unmentioned, Record::Deleted) => {
+                    // The item isn't mentioned locally, but is deleted
+                    // remotely. This is the case for a first sync, and will
+                    // change if we store tombstones (bug 1343103). For now,
+                    // we only flag the tombstone as remotely merged.
+                    ops.set_remote_merged.push(SetRemoteMerged(guid));
+                }
+                (Record::Exists(local_node), Record::Unmentioned) => {
+                    // The item exists locally, but isn't mentioned remotely.
+                    // Delete it from the local tree, but don't upload a
+                    // tombstone or flag it as remotely merged, since it's not
+                    // on the server. This is the case for invalid local-only
+                    // items.
+                    ops.delete_local_items.push(DeleteLocalItem(local_node));
+                }
+                (Record::Exists(local_node), Record::Exists(remote_node)) => {
+                    // The item exists on both sides, and is flagged for
+                    // deletion. This is the case for non-syncable and invalid
+                    // items. Delete the item from the local tree and upload a
+                    // tombstone, but don't flag it as remotely merged; storing
+                    // the uploaded tombstone takes care of that.
+                    ops.delete_local_items.push(DeleteLocalItem(local_node));
+                    ops.insert_local_tombstones
+                        .push(InsertLocalTombstone(remote_node));
+                    ops.upload_tombstones
+                        .push(UploadTombstone(&remote_node.item().guid));
+                }
+                (Record::Exists(local_node), Record::Deleted) => {
+                    // The item exists locally, and is deleted remotely. This is
+                    // the most common case for incoming deletions. Delete the
+                    // item from the local tree, and flag the tombstone as
+                    // merged.
+                    ops.delete_local_items.push(DeleteLocalItem(local_node));
+                    ops.set_remote_merged.push(SetRemoteMerged(guid));
+                }
+                (Record::Deleted, Record::Unmentioned) => {
+                    // Uploading a tombstone for a nonexistent item. Nothing
+                    // to do here.
```

Comment:
We might as well upload it. We don't have a "remove tombstone" op, and not uploading here means there's a difference between what we upload, and what's in `moz_bookmarks_deleted`. Ditto for `(Record::Deleted, Record::Deleted)` below.